source/gui/settingsDialogs.py

```diff
@@ -5443,3 +5523,197 @@ def onFilterEditTextChange(self, evt):
+		self,
+		parent: wx.Window,
+		title: str,
+		configPath: Iterable[str],
+		helpId: str | None = None,
+		urlTransformer: Callable[[str], str] = lambda url: url,
+		*args,
+		**kwargs,
+	):
+		"""Customisable dialog for requesting a URL from the user.
+
+		:param parent: Parent window of this dialog.
+		:param title: Title of this dialog.
+		:param configPath: Where in the config the URL is to be stored.
+		:param helpId: Anchor of the user guide section for this dialog, defaults to None
+		:param urlTransformer: Function to transform the given URL into something usable, eg by adding required query parameters. Defaults to the identity function.
+		:raises ValueError: If no config path is given.
+		"""
+		if not configPath or len(configPath) < 1:
+			raise ValueError("Config path not provided.")
+		self.title = title
+		self.helpId = helpId
+		self._configPath = configPath
+		self._urlTransformer = urlTransformer
+		super().__init__(parent, *args, **kwargs)
+
+	def makeSettings(self, settingsSizer):
+		settingsSizerHelper = guiHelper.BoxSizerHelper(self, sizer=settingsSizer)
+		self._urlControl = urlControl = settingsSizerHelper.addLabeledControl(
+			# Translators: The label of a text box asking the user for a URL.
+			# The purpose of this text box will be explained elsewhere in the user interface.
+			_("&URL:"),
+			wx.TextCtrl,
+			size=(250, -1),
+		)
+		self.bindHelpEvent("UpdateMirrorURL", urlControl)
+		self._testButton = testButton = wx.Button(
+			self,
+			# Translators: A button in a dialog which allows the user to test a URL that they have entered.
+			label=_("&Test..."),
+		)
+		self.bindHelpEvent("UpdateMirrorTest", testButton)
+		urlControlsSizerHelper = guiHelper.BoxSizerHelper(self, sizer=urlControl.GetContainingSizer())
+		urlControlsSizerHelper.addItem(testButton)
+		testButton.Bind(wx.EVT_BUTTON, self._onTest)
+		# Order of operations is important here.
+		urlControl.Bind(wx.EVT_TEXT, self._onTextChange)
+		self._url = self._getFromConfig()
+
+	def postInit(self):
+		# Ensure that focus is on the URL text box.
+		self._urlControl.SetFocus()
+
+	def onOk(self, evt: wx.CommandEvent):
+		self._normalize()
+		if self._url == self._getFromConfig():
+			shouldSave = False
+		elif self._testStatus != SetURLDialog._URLTestStatus.PASSED:
+			ret = gui.messageBox(
+				_(
+					# Translators: Message shown to users when saving a potentially invalid URL to NVDA's settings.
+					"The URL you have entered failed the connection test. Are you sure you want to save it anyway?",
+				)
+				if self._testStatus == SetURLDialog._URLTestStatus.FAILED
+				else _(
+					# Translators: Message shown to users when saving an untested URL to NVDA's settings.
+					"The URL you have entered has not been tested. Are you sure you want to save it without attempting to connect to it first?",
+				),
+				# Translators: The title of a dialog.
+				_("Warning"),
+				wx.OK | wx.CANCEL | wx.CANCEL_DEFAULT | wx.ICON_WARNING,
+				self,
+			)
```

Comment:
Should this have a "don't show this warning again" option?

Reply:
Surely not. We do not expect this value to be changed often so the confirmation dialog should not bother often either.
Adding a don't show again option would just make the GUI more complex here.

The don't show again are meant to be used with dialogs that appear quite regularly, e.g. opening the Add-on Store, enabling the screen curtain, etc.

Reply:
That was my thinking too, so I'm glad we're on the same page. :)